lib/core/src/Cardano/Wallet/Primitive/Types/Tx.hs

```diff
@@ -392,11 +416,92 @@ instance Buildable a => Buildable (WithDirection a) where
         <> (case d of; Incoming -> "+"; Outgoing -> "-")
         <> build a
 
--- | @SealedTx@ is a signed and serialised transaction that is ready to be
--- submitted to the node.
-newtype SealedTx = SealedTx { getSealedTx :: ByteString }
-    deriving stock (Show, Eq, Generic)
-    deriving newtype (ByteArrayAccess, NFData)
+-- | 'SealedTx' is a possibly signed and serialised transaction that is ready to
+-- be submitted to the node.
+--
+-- Construct it with either 'sealedTxFromCardano' or 'sealedTxFromBytes'.
+data SealedTx = SealedTx
+    { cardanoTx :: InAnyCardanoEra Cardano.Tx
+    , serialisedTx :: ByteString
+    } deriving stock (Generic)
+
+instance Show SealedTx where
+    show (SealedTx (InAnyCardanoEra _era tx) _) = show tx
+
+instance Eq SealedTx where
+    SealedTx (InAnyCardanoEra e1 _) a == SealedTx (InAnyCardanoEra e2 _) b =
+        case testEquality e1 e2 of
+            Just Refl -> a == b
+            Nothing -> False
+
+instance NFData SealedTx where
+    rnf = rnf . show  -- fixme: temp fix
+
+-- | Construct a 'SealedTx' from a "Cardano.Api" transaction.
+sealedTxFromCardano :: InAnyCardanoEra Cardano.Tx -> SealedTx
+sealedTxFromCardano tx = SealedTx tx (cardanoTxToBytes tx)
+  where
+    cardanoTxToBytes :: InAnyCardanoEra Cardano.Tx -> ByteString
+    cardanoTxToBytes (InAnyCardanoEra _era tx') = Cardano.serialiseToCBOR tx'
+
+-- | Deserialise a Cardano transaction. The transaction can be in the format of
+-- any era. This function will try the most recent era first, then
+-- previous eras until 'ByronEra'.
+cardanoTxFromBytes
+    :: AnyCardanoEra -- ^ Most recent era
+    -> ByteString -- ^ Serialised transaction
+    -> Either DecoderError (InAnyCardanoEra Cardano.Tx)
+cardanoTxFromBytes maxEra bs = asum $ map snd $ filter (withinEra maxEra . fst)
+    [ (anyCardanoEra AlonzoEra, InAnyCardanoEra AlonzoEra <$> deserialiseFromCBOR (Cardano.AsTx Cardano.AsAlonzoEra) bs)
+    , (anyCardanoEra MaryEra, InAnyCardanoEra MaryEra <$> deserialiseFromCBOR (Cardano.AsTx Cardano.AsMaryEra) bs)
+    , (anyCardanoEra AllegraEra, InAnyCardanoEra AllegraEra <$> deserialiseFromCBOR (Cardano.AsTx Cardano.AsAllegraEra) bs)
+    , (anyCardanoEra ShelleyEra, InAnyCardanoEra ShelleyEra <$> deserialiseFromCBOR (Cardano.AsTx Cardano.AsShelleyEra) bs)
+    , (anyCardanoEra ByronEra, InAnyCardanoEra ByronEra <$> deserialiseFromCBOR (Cardano.AsTx Cardano.AsByronEra) bs)
+    ]
+  where
```

Comment:
Suggestion: extract out the shared logic into a separate function:

```suggestion
cardanoTxFromBytes maxEra bs = asum $ map snd $ filter (withinEra maxEra . fst)
    [ deserialise AlonzoEra  Cardano.AsAlonzoEra
    , deserialise MaryEra    Cardano.AsMaryEra
    , deserialise AllegraEra Cardano.AsAllegraEra
    , deserialise ShelleyEra Cardano.AsShelleyEra
    , deserialise ByronEra   Cardano.AsByronEra
    ]
  where
    deserialise
        :: forall era. Cardano.IsCardanoEra era
        => CardanoEra era
        -> Cardano.AsType era
        -> (AnyCardanoEra, Either DecoderError (InAnyCardanoEra Cardano.Tx))
    deserialise era asEra =
        ( anyCardanoEra era
        , InAnyCardanoEra era <$> deserialiseFromCBOR (Cardano.AsTx asEra) bs
        )
```

Reply:
Yep good idea, thanks.